From bc6ed8fe0b2691b92f0d508d73e1b903aee5fd63 Mon Sep 17 00:00:00 2001
From: Kemeng Shi <shikemeng@huawei.com>
Date: Thu, 27 May 2021 14:42:55 +0800
Subject: [PATCH 42/50] etmemd: fix memleak and clean code

1.detach cslide main thread to release resource when exit
2.clean code

Signed-off-by: Kemeng Shi <shikemeng@huawei.com>
---
 src/etmemd_src/etmemd_cslide.c | 6 +++++-
 src/etmemd_src/etmemd_task.c   | 7 ++++++-
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/src/etmemd_src/etmemd_cslide.c b/src/etmemd_src/etmemd_cslide.c
index d1dd06c..ad3eff8 100644
--- a/src/etmemd_src/etmemd_cslide.c
+++ b/src/etmemd_src/etmemd_cslide.c
@@ -1606,6 +1606,10 @@ static void *cslide_main(void *arg)
     struct cslide_eng_params *eng_params = (struct cslide_eng_params *)arg;
     struct sys_mem *mem = NULL;
 
+    // only invalid pthread id or deatch more than once will cause error
+    // so no need to check return value of pthread_detach
+    (void)pthread_detach(pthread_self());
+
     while (true) {
         factory_update_pid_params(&eng_params->factory);
         if (eng_params->finish) {
@@ -1712,7 +1716,7 @@ static char *get_time_stamp(time_t *t)
         return NULL;
     }
 
-    ts = asctime(localtime(t));
+    ts = asctime(lt);
     if (ts == NULL) {
         etmemd_log(ETMEMD_LOG_ERR, "get asctime fail\n");
         return NULL;
diff --git a/src/etmemd_src/etmemd_task.c b/src/etmemd_src/etmemd_task.c
index 5aa693b..618245e 100644
--- a/src/etmemd_src/etmemd_task.c
+++ b/src/etmemd_src/etmemd_task.c
@@ -58,7 +58,12 @@ static int get_pid_through_pipe(char *arg_pid[], const int *pipefd)
             return -1;
         }
 
-        execve(arg_pid[0], arg_pid, NULL);
+        if (execve(arg_pid[0], arg_pid, NULL) == -1) {
+            etmemd_log(ETMEMD_LOG_ERR, "execve %s fail with %s.\n", arg_pid[0], strerror(errno));
+            close(pipefd[1]);
+            return -1;
+        }
+
         if (fflush(stdout) != 0) {
             etmemd_log(ETMEMD_LOG_ERR, "fflush execve stdout fail.\n");
             close(pipefd[1]);
-- 
2.27.0

